controllers/suite_test.go

```diff
@@ -236,110 +138,73 @@ var _ = BeforeEach(func() {
-				},
-				{
-					Image: "car",
-					Name:  "ctn3",
-				},
-			},
-		},
-	}
-	targetPodD = &corev1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "far",
-			Namespace: "default",
-			Labels: map[string]string{
-				"foo": "bar",
-			},
-		},
-		Spec: corev1.PodSpec{
-			Containers: []corev1.Container{
-				{
-					Image: "far",
-					Name:  "ctn1",
-				},
-				{
-					Image: "far",
-					Name:  "ctn2",
-				},
-				{
-					Image: "far",
-					Name:  "ctn3",
-				},
 			},
 		},
 	}
 
-	By("Creating target pod")
-	err := k8sClient.Create(context.Background(), targetPodA)
-	if apierrors.IsInvalid(err) {
-		logf.Log.Error(err, "failed to create object, got an invalid object error")
-		return
-	}
-	Expect(err).NotTo(HaveOccurred())
+	By("Creating target pods")
+	Expect(k8sClient.Create(context.Background(), targetPodA)).ToNot(HaveOccurred())
+	Expect(k8sClient.Create(context.Background(), targetPodB)).ToNot(HaveOccurred())
 
-	err = k8sClient.Create(context.Background(), targetPodB)
-	if apierrors.IsInvalid(err) {
-		logf.Log.Error(err, "failed to create object, got an invalid object error")
-		return
-	}
-	Expect(err).NotTo(HaveOccurred())
+	By("Waiting for target pods to be ready")
+	Eventually(func() error {
+		running, err := podsAreRunning(targetPodA, targetPodB)
+		if err != nil {
+			return err
+		}
 
-	err = k8sClient.Create(context.Background(), targetPodC)
-	if apierrors.IsInvalid(err) {
-		logf.Log.Error(err, "failed to create object, got an invalid object error")
-		return
-	}
-	Expect(err).NotTo(HaveOccurred())
+		if !running {
+			return fmt.Errorf("target pods are not running")
+		}
 
-	err = k8sClient.Create(context.Background(), targetPodD)
-	if apierrors.IsInvalid(err) {
-		logf.Log.Error(err, "failed to create object, got an invalid object error")
-		return
-	}
-	Expect(err).NotTo(HaveOccurred())
-})
+		return nil
+	}, timeout).Should(Succeed())
+
+	close(done)
+}, 60)
 
 var _ = AfterSuite(func() {
 	By("tearing down the test environment")
-	err := testEnv.Stop()
-	Expect(err).ToNot(HaveOccurred())
-})
 
-var _ = AfterEach(func() {
 	_ = k8sClient.Delete(context.Background(), targetPodA)
 	_ = k8sClient.Delete(context.Background(), targetPodB)
-	_ = k8sClient.Delete(context.Background(), targetPodC)
-	_ = k8sClient.Delete(context.Background(), targetPodD)
+
+	err := testEnv.Stop()
+	Expect(err).ToNot(HaveOccurred())
 })
+
+// podsAreRunning returns true when all the given pods have all their containers running
+func podsAreRunning(pods ...*corev1.Pod) (bool, error) {
+	for _, pod := range pods {
+		var p corev1.Pod
+
+		// retrieve pod
+		if err := k8sClient.Get(context.Background(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, &p); err != nil {
+			return false, fmt.Errorf("error getting pod: %w", err)
+		}
+
+		// check the pod phase
+		if p.Status.Phase != corev1.PodRunning {
+			return false, nil
+		}
+
+		// check the pod containers statuses (phase can be running while all pods are not running)
```

Comment:
should this say "while all containers are not running"?

Reply:
Indeed, thanks